### PR TITLE
journal/cat: allow connecting output to specific journal namespace

### PR DESCRIPTION
**Pull request by YHNdnzj**

### CODE REVIEW

File: src/boot/bless-boot.c [hunk 1]
```diff
@@ -370,22 +363,22 @@ static int verb_status(int argc, char *argv[], void *userdata) {
                         return log_error_errno(errno, "Failed to open $BOOT partition '%s': %m", *p);
                 }
 
-                if (faccessat(fd, skip_slash(path), F_OK, 0) >= 0) {
+                if (faccessat(fd, skip_slash_or_dot(path), F_OK, 0) >= 0) {
```
Comment (yuwata) [inline src/boot/bless-boot.c]: Hmm, I do not think it is necessary to drop `./` in this case and below. Maybe, `skip_leading_chars(path, "/")` is enough?

File: src/libsystemd/sd-journal/journal-internal.h [hunk 2]
```diff
@@ -151,3 +152,15 @@ static inline bool JOURNAL_ERRNO_IS_UNAVAILABLE_FIELD(int r) {
                       E2BIG,            /* Field too large for pointer width */
                       EPROTONOSUPPORT); /* Unsupported compression */
 }
+
+#define journal_stream_path(log_namespace)                                              \
```
Comment (yuwata) [inline src/libsystemd/sd-journal/journal-internal.h]: Nah, please implement it as a normal function with strjoin().

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/boot/bless-boot.c (hunk 1 above)
Comment (YHNdnzj) [reply src/boot/bless-boot.c]: Hmm, we pass these to `_at` functions, so `./` has no effect either way?

Comment (yuwata) [reply src/boot/bless-boot.c]: Right. I do not say using skip_slash_or_dot() is wrong, but that is not necessary.

File: src/libsystemd/sd-journal/journal-internal.h (hunk 2 above)
Comment (YHNdnzj) [reply src/libsystemd/sd-journal/journal-internal.h]: Well, that way we'd lose the "async signal safe" attribute. I'm not sure if that's acceptable?

Comment (YHNdnzj) [reply src/libsystemd/sd-journal/journal-internal.h]: I'm not too concerned that people would actually call these in signal handlers, though...

Comment (YHNdnzj) [reply src/libsystemd/sd-journal/journal-internal.h]: Hmm, so obviously people do rely on this: https://github.com/systemd/systemd/issues/7912.

Comment (yuwata) [reply src/libsystemd/sd-journal/journal-internal.h]: Hmm, then please add comment why this is a macro.

**Comment by yuwata** [comment]
Please add a test case for --namespace option.

**Review by yuwata — APPROVED** [review]
LGTM.

File: src/libsystemd/sd-journal/journal-send.h [hunk 3]
```diff
@@ -2,6 +2,23 @@
...
+
+/* We declare sd_journal_stream_fd() as async-signal-safe. So instead of strjoin(), which calls malloc()
+ * internally, use a macro + alloca(). */
+#define journal_stream_path(log_namespace)                                              \
+        ({                                                                              \
+                const char *_ns = (log_namespace), *_ret;                               \
+                if (!_ns)                                                               \
+                        _ret = "/run/systemd/journal/stdout";                           \
+                else if (log_namespace_name_valid(_ns))                                 \
+                        _ret = strjoina("/run/systemd/journal.", _ns, "/stdout");       \
```
Comment (keszybz) [inline src/libsystemd/sd-journal/journal-send.h:20]: The space allocated via alloca remains valid until the end of the calling function, so this seems OK to do here.

**Review by keszybz — APPROVED** [review]
LGTM.

**Comment by poettering** [comment]
lgtm

File: src/journal/cat.c [hunk 4]
```diff
@@ -44,6 +45,7 @@ static int help(void) {
                "  -p --priority=PRIORITY         Set priority value (0..7)\n"
                "     --stderr-priority=PRIORITY  Set priority value (0..7) used for stderr\n"
                "     --level-prefix=BOOL         Control whether level prefix shall be parsed\n"
+               "     --namespace=NAMESPACE       Connect to specified journal namespace\n"
```
Comment (bluca) [inline src/journal/cat.c:48]: @YHNdnzj could you please add bash completion for this as a follow-up?

Comment (YHNdnzj) [reply src/journal/cat.c:48]: -> #31784

File: src/libsystemd/sd-journal/journal-send.h [hunk 5]
```diff
@@ -2,6 +2,23 @@
...
+#include <stddef.h>
+
+#include "syslog-util.h"
 
 int journal_fd_nonblock(bool nonblock);
 void close_journal_fd(void);
+
+/* We declare sd_journal_stream_fd() as async-signal-safe. So instead of strjoin(), which calls malloc()
+ * internally, use a macro + alloca(). */
+#define journal_stream_path(log_namespace)                                              \
```
Comment (poettering) [inline src/libsystemd/sd-journal/journal-send.h:14]: frankly, I'd just make this a regular functin, returning heap memory.

File: src/journal/cat.c [hunk 6]
```diff
@@ -117,6 +121,13 @@ static int parse_argv(int argc, char *argv[]) {
                                 return r;
                         break;
 
+                case ARG_NAMESPACE:
+                        if (isempty(optarg))
+                                arg_namespace = NULL;
+                        else
+                                arg_namespace = optarg;
```
Comment (poettering) [inline src/journal/cat.c:128]: arg_namespace = empty_to_null(optarg);

**Comment by poettering** [comment]
some minor post-merge comments above

File: src/libsystemd/sd-journal/journal-send.h (hunk 5 above)
Comment (YHNdnzj) [reply src/libsystemd/sd-journal/journal-send.h:14]: We can't without breaking some documented guarantees? As the comment stated.

File: src/journal/cat.c (hunk 6 above)
Comment (YHNdnzj) [reply src/journal/cat.c:128]: Already fixed in 80fcdb731

**Pull request merged**